Ad Blocking Recovery onboarding completion

nfmohit opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #6966 

## Relevant technical choices

This PR aims to complete the Ad Blocking Recovery onboarding screen.

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

nfmohit commented on assets/js/modules/adsense/components/settings/AdBlockingRecoveryApp/index.js:
```diff
@@ -0,0 +1,223 @@
+/**
```
Moved the component to its own folder because it was getting too large, hence I decided to have each step contents in their own sub-components.

nfmohit commented on includes/Modules/AdSense/Settings.php:
```diff
@@ -229,6 +229,7 @@ protected function get_sanitize_callback() {
 						! in_array(
 							$option['adBlockingRecoverySetupStatus'],
 							array(
+								'',
```
Allowed this to be an empty string so that the cancellation flow works as described in the ACs.

aaemnnosttv replied:
Sounds reasonable; I would think all of the default values should be acceptable.

aaemnnosttv commented on assets/js/modules/adsense/components/settings/AdBlockingRecoveryApp/index.js:
```diff
@@ -0,0 +1,223 @@
+				<Grid>
+					<Row>
+						<Cell size={ 12 }>
+							<Layout rounded>
```
This isn't an overly-large component on the whole, but this `Layout` and within seems like a good place to extract one which would leave the top-level index here as the skeleton for the page. Maybe we'd call this the `SetupMain` of the app?

nfmohit replied:
I agree. Updated the structure accordingly, thanks!

aaemnnosttv commented on assets/js/modules/adsense/components/settings/AdBlockingRecoveryApp/index.js:
```diff
@@ -0,0 +1,223 @@
+									</Row>
+								</Grid>
+
+								<Grid className="googlesitekit-ad-blocking-recovery__content">
+									<Row>
+										<Cell mdSize={ 6 } lgSize={ 8 }>
```
This could also be another component for `Content` perhaps, which would essentially just be the grid, but also include the graphic so that `children` would essentially just be the `Stepper` here.

nfmohit replied:
Updated accordingly, thank you!

aaemnnosttv commented on assets/js/modules/adsense/components/settings/AdBlockingRecoveryApp/index.stories.js:
```diff
@@ -0,0 +1,114 @@
+					.dispatch( MODULES_ADSENSE )
+					.receiveIsAdBlockerActive( false );
+
+				registry.dispatch( CORE_SITE ).receiveSiteInfo( {
+					adminURL: 'http://example.com/wp-admin/',
+				} );
```
Please use `provideSiteInfo` instead. This uses the same `adminURL` by default internally, so providing one explicitly isn't necessary but can be useful if a different value is needed, or to make the test more explicit when the value is relevant.

nfmohit replied:
Excellent catch, updated, thank you!

aaemnnosttv commented on assets/js/modules/adsense/components/settings/AdBlockingRecoveryApp/steps/PlaceTagsStep.js:
```diff
@@ -0,0 +1,132 @@
+		const { error } = await saveSettings();
+
+		if ( ! error ) {
+			setActiveStep( 1 );
```
1 is a bit magical here. Does that mean the first step or the second? Or should it be a specific step, regardless of the order. Let's introduce an enum for referencing specific steps.

Also, what happens when there is an error?

nfmohit replied:
> 1 is a bit magical here. Does that mean the first step or the second? Or should it be a specific step, regardless of the order. Let's introduce an enum for referencing specific steps.

Introduced enum to reference steps.

> Also, what happens when there is an error?

Updated the components to display the error message if there is an error.

Thank you!

aaemnnosttv commented on assets/sass/components/adsense/_googlesitekit-adsense-ad-blocking-recovery.scss:
```diff
@@ -147,4 +147,60 @@
+	.googlesitekit-ad-blocking-recovery__step-place-tags {
+
+		.mdc-checkbox {
+			margin-left: -11px;
```
This is a bit odd however I see the same here https://github.com/google/site-kit-wp/blob/471e3665756f1d5589a71d45e471e19c5a8fd7cd/assets/sass/components/global/_googlesitekit-view-only-menu.scss#L119-L121

Are we missing this style on the checkbox itself? If not, but we need this in some cases and not others, it should probably be handled by a class to avoid duplicating this every time.

nfmohit replied:
I have addressed this in a slightly big commit - https://github.com/google/site-kit-wp/pull/7183/commits/97561d2f1707060439a70b452c6854bf1cdaa293.

To summarise, I have introduced a new prop to the `Checkbox` and `Radio` components - `alignLeft`, which under the hood applies the negative margin to compensate for the additional padding of the components. I have also replaced the existing negative margin usages to use this prop instead.

Please let me know if this is acceptable. Thank you!

aaemnnosttv commented on assets/sass/components/adsense/_googlesitekit-adsense-ad-blocking-recovery.scss:
```diff
@@ -147,4 +147,60 @@
+		p.googlesitekit-ad-blocking-recovery__error-protection-tag-info {
+			font-size: $fs-body-md;
+			line-height: $lh-body-md;
+			margin: 0 0 20px 34px;
+		}
+	}
+
+	.googlesitekit-ad-blocking-recovery__step-create-message {
+
+		.googlesitekit-ad-blocking-recovery__create-message-footer-actions {
+			align-items: center;
+			display: flex;
+			flex-wrap: wrap;
+			gap: 26px;
```
Similar to above, are there not any tokens we can use for these literal px values? Things like margin and gap should be standard so I wouldn't expect we'd need exceptions. If we did, let's document why those are.

nfmohit replied:
Great catch, I've updated it to use tokens wherever possible, thank you!

aaemnnosttv commented on assets/sass/components/adsense/_googlesitekit-adsense-ad-blocking-recovery.scss:
```diff
@@ -147,4 +147,60 @@
+
+		.googlesitekit-ad-blocking-recovery__create-message-footer {
+			display: inline-block;
+			margin-top: 12px;
+		}
+
+		p.googlesitekit-ad-blocking-recovery__create-message-footer-note {
+			font-size: $fs-body-sm;
+			letter-spacing: $ls-xs;
+			line-height: $lh-body-sm;
+			margin-top: 24px;
```
Same here

nfmohit replied:
Updated, thank you!

aaemnnosttv commented on assets/js/modules/adsense/components/settings/AdBlockingRecoveryApp/steps/PlaceTagsStep.js:
```diff
@@ -0,0 +1,132 @@
+		await syncAdBlockingRecoveryTags();
+
+		setAdBlockingRecoverySetupStatus(
+			AD_BLOCKING_RECOVERY_SETUP_STATUS_TAG_PLACED
```
Similar to another comment, for constants which are all related like this (valid set of values) an enum would be a better fit than a list of constants that follow a naming convention.

nfmohit replied:
Addressed, thank you!

aaemnnosttv commented on assets/js/modules/adsense/components/settings/AdBlockingRecoveryApp/index.stories.js:
```diff
@@ -0,0 +1,114 @@
+	decorators: [
+		( Story, { args } ) => {
+			const setupRegistry = ( registry ) => {
+				provideModules( registry, [
+					{
+						active: true,
+						connected: true,
+						slug: 'adsense',
+					},
+				] );
```
This makes sense, but is it actually required in this context?

nfmohit replied:
This actually preexisted before I picked up this issue, but you're correct, it isn't required in this context. Removed, thanks!

aaemnnosttv reviewed (changes requested):
Thanks @nfmohit ! This looks solid overall, just a few questions/suggestions to address. Nice work!

nfmohit commented:
> Thanks @nfmohit ! This looks solid overall, just a few questions/suggestions to address. Nice work!

Thank you very much for the kind and detailed review, @aaemnnosttv! I have addressed all the comments and updated the PR. Back to you for another look.

aaemnnosttv commented on assets/js/googlesitekit/components-gm2/Checkbox.js:
```diff
@@ -44,13 +44,15 @@ export default function Checkbox( props ) {
 		tabIndex,
 		onKeyDown,
 		loading,
+		alignLeft,
```
Do we also need to apply this to the gm3 checkbox?

nfmohit replied:
Applied, thank you!

aaemnnosttv commented on assets/sass/components/adsense/_googlesitekit-adsense-ad-blocking-recovery.scss:
```diff
@@ -147,4 +147,57 @@
+			align-items: center;
+			display: flex;
+			flex-wrap: wrap;
+			gap: $grid-gap-desktop + 2px;
```
Why the extra 2px? 🤔

nfmohit replied:
According to the Figma designs, the buttons have a gap of 26px in between:

![image](https://github.com/google/site-kit-wp/assets/20284937/cb1a11b1-5b65-4a48-9970-770a3c028b99)

Since `$grid-gap-desktop` is set to 24px, I added 2px to match the designs. Should I remove it?

aaemnnosttv replied:
Ah, I see now. So the secondary choice there is actually a button in the design, but we use a link. The two have a 10px gap between.
<img width="812" alt="image" src="https://github.com/google/site-kit-wp/assets/1621608/e5b1bc3a-bb40-412f-ac52-8dca75b2b4e2">

We probably don't have a great solution for this currently I don't think. Ideally we'd have maybe an `ActionsRow` component in the future to have a consistent layout for this kind of thing but for now, using a `26` would be okay, I'd just add a comment as to where the value is coming from. This is something we should address throughout, so we don't need to get overly hung up on it here 👍

nfmohit replied:
Ah, I see, that makes sense. I have added a comment to explain where the extra 2px comes from. Thank you!

aaemnnosttv commented on assets/sass/vendor/_mdc-radio.scss:
```diff
@@ -30,4 +30,12 @@
+	// to the absolute left when required. The following style compensates for the
+	// padding and aligns it to the absolute left.
+	&.mdc-radio--align-left {
+		margin-left: -10px;
```
The comment here says 11px but this is only 10. Also, the file `assets/sass/components/user-input/googlesitekit-user-input-preview.scss` uses 10px for both checkbox and radios. It's fine if they should be different for accuracy, but just noting this difference from the previous CSS.

Is there a custom property that material uses for the padding that we should reference here instead?

nfmohit replied:
> The comment here says 11px but this is only 10. Also, the file `assets/sass/components/user-input/googlesitekit-user-input-preview.scss` uses 10px for both checkbox and radios. It's fine if they should be different for accuracy, but just noting this difference from the previous CSS.

Ah, I copy pasted the comment from Checkbox but mistakenly didn't update this detail, updated it now, thank you!

> Is there a custom property that material uses for the padding that we should reference here instead?

[According to the material codebase](https://github.com/material-components/material-components-web/blob/b994146f6d849bd80f280c147d64adae5b930a3c/packages/mdc-checkbox/_checkbox-theme.scss#L342), it seems to be the ripple size minus icon size divided by 2. Do you think there is something usable here?

aaemnnosttv replied:
Hmm that looks complicated 😄 There's probably a way to do it properly with a Material theme mixin but it's not worth spending that much time on.

nfmohit replied:
I agree 😄 Thank you!

aaemnnosttv commented on assets/js/modules/adsense/components/settings/AdBlockingRecoveryApp/index.stories.js:
```diff
@@ -0,0 +1,103 @@
+};
+
+export default {
+	title: 'Modules/AdSense/Settings/AdBlockingRecoveryApp',
+	decorators: [
```
Let's set the `padding` to 0 for these stories to avoid adding this gap since the stories here are for an "app" rather than an inner component. We should be able to do this for all of them at this part of the story definition.

<img width="466" alt="image" src="https://github.com/google/site-kit-wp/assets/1621608/766d5553-35f1-4ae8-9448-0022d09d02ea">

nfmohit replied:
Updated, thank you!

aaemnnosttv commented on assets/js/modules/adsense/components/settings/AdBlockingRecoveryApp/index.stories.js:
```diff
@@ -0,0 +1,103 @@
+	},
+};
+StepOne.scenario = {
+	label: 'Global/AdBlockingRecoveryApp/StepOne',
```
Nitpick: `StepOne` can be misinterpreted as a component here. The last segment should be more of a human-readable label of the scenario. This is consistent with what we have now so this is more of a note for a holistic change since there are many stories that use this now.

These should not however be categorized under "Global" but instead under `Modules/AdSense`.

nfmohit replied:
Updated the labels to be more human-readable, let me know if that looks good, thank you!

aaemnnosttv commented on assets/js/modules/adsense/components/settings/AdBlockingRecoveryApp/SetupMain.js:
```diff
@@ -0,0 +1,175 @@
+		const adBlockingRecoverySetupStatus =
+			select( MODULES_ADSENSE ).getAdBlockingRecoverySetupStatus();
+
+		switch ( adBlockingRecoverySetupStatus ) {
+			case '':
+				return ENUM_AD_BLOCKING_RECOVERY_SETUP_STEP.PLACE_TAGS;
+			case ENUM_AD_BLOCKING_RECOVERY_SETUP_STATUS.TAG_PLACED:
+				return ENUM_AD_BLOCKING_RECOVERY_SETUP_STEP.CREATE_MESSAGE;
+			case ENUM_AD_BLOCKING_RECOVERY_SETUP_STATUS.SETUP_CONFIRMED:
+				return ENUM_AD_BLOCKING_RECOVERY_SETUP_STEP.COMPLETE;
+		}
```
This mapping could be defined once at the top and then this could be simplified to `statusStepMap[status] || ENUM_AD_BLOCKING_RECOVERY_SETUP_STEP.PLACE_TAGS`

Not a big deal but it is a little hard to read as-is.

nfmohit replied:
Very good idea, updated, let me know if this looks better now, thank you!

nfmohit commented:
> Thanks @nfmohit ! A few final comments but otherwise this is great!

Addressed, back to you for another look, thank you @aaemnnosttv!

aaemnnosttv reviewed (approved):
Great, thanks @nfmohit !
